### PR TITLE
Translate data sqids

### LINKED ISSUE #29: Create parse element ids to labels functionality
## Is your feature request related to a problem? Please describe.
Analysts will want to have meaningful column names and items in their API results, but the raw response contains largely API IDs as default.

## Describe the solution you'd like
Use a data set's meta data to translate IDs in a retrieved data query result into assigned labels.

So steps would be:
- data retrieved using post_dataset or get_dataset
- function converts IDs to labels
- returns data frame with labels instead of ids

## Describe alternatives you've considered
Leave analysts to do it themselves. Just sounds a bit annoying and we may as well centralise it.

### PR DESCRIPTION
**Pull request by rmbielby**
# Brief overview of changes

Adding in some translation of SQIDs into understandable labels and codes based on a data set's meta data.

## Why are these changes being made?

SQIDs don't make any sense to end users, they're going to want them translating and we can save them the effort.

## Detailed description of changes

I've added some parsing functions to handle the different types of objects. These are called in by `parse_api_dataset()`. Basic structure is along the lines:
- User runs `query_dataset()`, which then runs `post_dataset()` or `get_datatset()`
  - As they retrieve pages of data, `post_dataset()` and `get_datatset()` translate the IDs on the fly using `parse_api_dataset()`
    - `parse_api_dataset()` then calls the following sub-functions:
      - `parse_sqids_locations()`
      - `parse_sqids_filters()`
      - `parse_sqids_indicators()`
     
## Additional information for reviewers

Related to discussion earlier, this isn't great for performance for large (paged) data sets as it runs this for every page. It would ultimately be better to run at the end of the process, but there's some not-insignificant rejigging in order to do that. Happy to look at that, but might be something for a separate issue after the initial bit of internal testing has started.

## Issue ticket number/s and link

Resolves #29

### CODE REVIEW

File: R/parse_sqids.R [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+      dplyr::left_join(lookup, by = column_sqid) |>
+      dplyr::select(-dplyr::all_of(column_sqid))
+  }
+  return(filters)
+}
+
+#' Parse IDs in a set of indicators
+#'
+#' @description
+#' The API uses unique IDs (sqids) to identify each filter column and its contents (filter items).
```
Comment (cjrace) [inline R/parse_sqids.R:97]: Filter description instead of indicators?

**Review by cjrace — CHANGES_REQUESTED** [review]
Few comments inline in the code, and then some bigger ones, which you would probably have been expecting, happy for these to go into separate issues to pick up at a later point if they're not quick to add in this PR:

1. We should add parsing for time, including a lookup data set for codes to identifiers
2. Locations parsing needs more refinement, would suggest a lookup data set for geographic level to code, including the prefix for each level's locations columns and also special handling for new_la_code / old_la_code and other less standard locations (RSC regions come to mind)
3. Should we give an option for whether or not to parse the output from `query_dataset()`? By default yes we should, though might be nice to have the option, will also be a good way to demonstrate how the IDs work by showing both versions as examples to users
4. Performance feels like it's going very backwards here, will definitely want a rewrite at a later point, though agree it is best to just get something out for testing and then optimise once we're more certain this will definitely be the approach we're running with

`query_dataset(example_id(), indicators = c("E40qF", "QuVwb", "9Aw4v"))`

Microbenchmarking before:
     min      lq     mean median       uq     max neval
 72.3144 75.9789 82.20526 77.926 82.88525 251.242   100

After
      min       lq    mean   median       uq      max neval
 184.6963 193.3727 203.332 199.0265 209.6245 343.5782   100

6. Very minor but mildly painful inconsistency in `query_dataset()` docs I spotted by accident...
image

File: R/examples.R [hunk 2]
```diff
@@ -94,6 +93,31 @@ example_id <- function(
   }
 }
 
+#' Title
```
[the remark this answers is not in the document]
Comment (rmbielby) [reply R/examples.R:96]: Updated

File: R/examples.R [hunk 3]
```diff
@@ -135,13 +159,13 @@ example_geography_query <- function(level = "nat_yorks") {
         return_level = c("NAT", "REG"),
         search_level = c("NAT", "REG"),
         identifier_type = c("code", "code"),
-        identifier = c("E92000001", "E12000002")
+        identifier = c("E92000001", "E12000003")
```
[the remark this answers is not in the document]
Comment (rmbielby) [reply R/examples.R:162]: Got to get that one right!

File: R/parse_sqids.R [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    dplyr::filter(!!rlang::sym("geographic_level_code") %in% names(locations)) |>
+    dplyr::rename(name = "label")
+  for (level in names(locations)) {
+    locations <- locations |>
+      dplyr::rename("item_id" = !!rlang::sym(level)) |>
+      dplyr::left_join(
+        lookup |>
+          dplyr::filter(!!rlang::sym("geographic_level_code") == level) |>
+          dplyr::select(-dplyr::all_of(c("geographic_level_code", "geographic_level"))) |>
+          dplyr::rename_with(~ paste0(tolower(level), "_", .x), !matches("item_id")),
```
[the remark this answers is not in the document]
Comment (rmbielby) [reply R/parse_sqids.R:30]: Yep, fixed now.

File: R/parse_sqids.R (hunk 1 above)
Comment (rmbielby) [reply R/parse_sqids.R:97]: Fixed

File: _pkgdown.yml [hunk 5]
```diff
@@ -35,6 +35,12 @@ reference:
   - get_dataset
   - post_dataset
   - parse_api_dataset
+  - starts_with("parse_sqid")
+
+- title: Generate example IDs and data
```
[the remark this answers is not in the document]
Comment (rmbielby) [reply _pkgdown.yml:40]: Removed the duplicate

[the remark this answers is not in the document]
Comment (cjrace) [reply _pkgdown.yml:40]: Think you missed this, still looks duplicated locally 
image

File: NAMESPACE [hunk 6]
```diff
@@ -34,3 +38,4 @@ export(validate_ees_id)
 export(validate_page_size)
 export(validate_time_periods)
 export(warning_max_pages)
+importFrom(data.table,":=")
```
[the remark this answers is not in the document]
Comment (rmbielby) [reply NAMESPACE:41]: I'm thinking of leaving this for now as a) I feel like lintr sometimes keeps the warnings on some things after you've fixed them and b) if it is still a genuine warning, then I'll be revisiting this whole bit of code when I come to doing it more sensibly with performance in mind.

**Review by cjrace — APPROVED** [review]

**Pull request merged**